R/zchunk_L111.nonghg_en_R_S_T_Y.R

```diff
@@ -1,111 +1,287 @@
+    L111.nonghg_tg_R_en_Sedgar_Yh.mlt %>%
+      filter(EDGAR_agg_sector %in% c("trn_intl_ship", "trn_intl_air")) ->
+      L111.nonghg_tg_R_en_Si_F_Yh_intl.mlt
+
+    L111.nonghg_tg_R_en_Sedgar_Yh.mlt %>%
+      filter(!(EDGAR_agg_sector %in% c("trn_intl_ship", "trn_intl_air"))) ->
+      L111.nonghg_tg_R_en_Si_F_Yh_dom.mlt
+
+    # Calculate total energy for international aviation emissions
+    L111.nonghg_tg_R_en_Si_F_Yh_intl.mlt %>%
+      group_by(EDGAR_agg_sector, Non.CO2, year) %>%
+      summarise(tot_energy = sum(energy)) %>%
+      ungroup ->
+      L111.en_Si_F_Yh_intl.mlt
+
+    # Compute international input emissions based on total emission and total emission shares
+    L111.nonghg_tg_R_en_Si_F_Yh_intl.mlt %>%
+      left_join(L111.en_Si_F_Yh_intl.mlt, by = c("EDGAR_agg_sector", "Non.CO2", "year")) %>%
+      left_join(L111_EDGAR_intl, by = c("EDGAR_agg_sector", "Non.CO2", "year")) %>%
+      # EDGAR emissions are in Gg, need to convert to Tg
+      mutate(emissions = (value * CONV_GG_TG) * (energy / tot_energy)) %>%
+      select(-tot_energy, -value) %>%
+      na.omit() %>%
+      # Add domestic emissions back in
+      bind_rows(L111.nonghg_tg_R_en_Si_F_Yh_dom.mlt) ->
+      L111.nonghg_tg_R_en_Si_F_Yh.mlt
+
+    # Map in final GCAM sector, technology, and driver type, and calculate energy totals (136-146)
+
+    L111.nonghg_tg_R_en_Si_F_Yh.mlt %>%
+      left_join_keep_first_only(GCAM_sector_tech, by = c("sector", "fuel", "technology")) %>%
+      # Sum emissions by region, non-CO2, sector, technology, and year for EDGAR historical years
+      group_by(GCAM_region_ID, Non.CO2, supplysector, subsector, stub.technology, year) %>%
+      summarise(value = sum(emissions)) %>%
+      ungroup %>%
+      filter(year %in% emissions.EDGAR_YEARS) ->
+      L111.nonghg_tg_R_en_S_F_Yh   # OUTPUT
+
+    # Compute non-ghg emission factors for GCAM sector, fuel, technology, and driver type (148-156)
+
+    L101.in_EJ_R_en_Si_F_Yh %>%
+      gather(year, value, matches(YEAR_PATTERN)) %>%
+      mutate(year = as.integer(year)) %>%
+      left_join_keep_first_only(select(GCAM_sector_tech, sector, fuel, technology, supplysector, subsector, stub.technology),
+                                by = c("sector", "fuel", "technology")) %>%
+      # need to filter here to replicate behavior of original aggregate()
+      filter(!is.na(supplysector), !is.na(subsector), !is.na(stub.technology)) %>%
+      group_by(GCAM_region_ID, supplysector, subsector, stub.technology, year) %>%
+      summarise(energy = sum(value)) %>%
+      ungroup ->
+      L101.in_EJ_R_en_S_F_Yh.mlt
+
+    # Calculate emission shares by energy totals
+
+    L111.nonghg_tg_R_en_S_F_Yh %>%
+      left_join(L101.in_EJ_R_en_S_F_Yh.mlt, by = c("GCAM_region_ID", "supplysector", "subsector", "stub.technology", "year")) %>%
+      # the old data system was doing a straight division of some _tiny_ numbers here, which made it impossible to replicate
```

Comment:
Why?  Are the tiny numbers coming from subtracting two nearly equal numbers?  If so, then that's a problem.  If, however, they come from multiplication or division, then it shouldn't be a problem, and any failure to replicate is probably coming from some other problem.

A good rule of thumb here is that any time you're performing an operation where catastrophic cancellation might be a problem (subtraction is a prime example), you should round to some appropriate number of significant figures _before_ the operation.  This ensures that the result will be flushed to zero if all of its significant digits canceled in the operation.

Reply:
I suspect we're going to want to audit the entire data system for this. At some point in the future.

Reply:
The issue comes up in the old data system in (energy module) LB1322.Fert.R, where a calculation sequence that should have yielded 0 was instead returning values on the order of ~1e-20. As @rplzzz suggested, it was addition/subtraction of values that were equal. Out of 3840 observations (32 region x 40 years x 3 fuels), something like 28 had this issue. AFAIK this bug has never mattered for these sectors (industrial energy use, industrial feedstocks) in GCAM because all calibration values are rounded to some reasonable number of digits. But the emissions module is using the numbers directly, with no rounding.
The issue behind the issue is that we have regions and time periods where our bottom-up estimate of natural gas consumption for fertilizer manufacturing exceeds the total industrial sector consumption of natural gas. This happens because our assignment of fertilizer production to fuel and feedstock source (i.e., share of gas vs oil vs coal) comes from an inventory year of 2004, when the vast majority of the production was from natural gas. However that wasn't the case in the early years of the time series (should have been more oil); we just don't have data on it.